[UnifiedPDF] Annotation tracking stops working when the mouse leaves the annotation.

sammygill opened the pull request:
#### e5936dcccb934ed15938c9a6374dac5964f230ab
<pre>
[UnifiedPDF] Annotation tracking stops working when the mouse leaves the annotation.
<a href="https://bugs.webkit.org/show_bug.cgi?id=270117">https://bugs.webkit.org/show_bug.cgi?id=270117</a>
<a href="https://rdar.apple.com/123123950">rdar://123123950</a>

Reviewed by Simon Fraser.

Currently annotation tracking begins in two scenarios:
1.) Mouse over of a texxt annotation
2.) Left click on other types of annotations

For the latter we finish tracking the annotation when the left mouse
button is released or if it is dragged off the tracked annotation. The
second case is incorrect, however, and we should contiue to keep track
of the annotation.

If the mouse is dragged off of the tracked annotation it should not be
highlighted anymore unless the mouse is dragged back onto it. If the
left mouse button is relased on the annotation then we want to perform
the associated action (e.g. activating a checkbox), but if it is
released off of the annotation then we should do nothing and just stop
tracking the annotation.

To accomplish the first part all we need to do is stop calling
finishTrackingAnnotation and instead call a new method:
updateTrackedAnnotation. All this new method does is update the
highlight of the tracked annotation but it does not modify the
annotation tracking state. For the second change, we need to pass in
the annotation under the mouse to finishTrackingAnnotation so the
annotation tracking state will only perform the associated action if
the annotation under the mouse is the annotation it is tracking.

One other small change that I made is to not change the cursor type when
the mouse is dragged off the annotation to make it consistent with
shipping.

* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::handleMouseEvent):
(WebKit::UnifiedPDFPlugin::updateTrackedAnnotation):
(WebKit::UnifiedPDFPlugin::finishTrackingAnnotation):
(WebKit::AnnotationTrackingState::finishAnnotationTracking):

Canonical link: <a href="https://commits.webkit.org/275362@main">https://commits.webkit.org/275362@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a9b38bc76682fb89b7bf73aef7a672a72aed5b58

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/41656 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/20670 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/44037 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/44226 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/37748 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/23786 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/18000 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/34420 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/42230 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/17584 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/35865 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/15086 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/15285 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/45605 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/37830 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/37204 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/40951 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/16471 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/13515 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/39367 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/18090 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/9328 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/18146 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/17734 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -1735,17 +1735,19 @@ static bool isContextMenuEvent(const WebMouseEvent& event)
-                notifyCursorChanged(toWebCoreCursorType({ }));
-                finishTrackingAnnotation(mouseEventType, mouseEventButton, RepaintRequirement::HoverOverlay);
+            if (RetainPtr trackedAnnotation = m_annotationTrackingState.trackedAnnotation()) {
+
```
You could remove this blank line

smfr commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -1792,7 +1794,9 @@ static bool isContextMenuEvent(const WebMouseEvent& event)
         case WebMouseEventButton::Left:
             if (RetainPtr trackedAnnotation = m_annotationTrackingState.trackedAnnotation(); trackedAnnotation && ![trackedAnnotation isKindOfClass:getPDFAnnotationTextWidgetClass()]) {
-                finishTrackingAnnotation(mouseEventType, mouseEventButton);
+
```
Ditto

smfr commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -1735,17 +1735,19 @@ static bool isContextMenuEvent(const WebMouseEvent& event)
+
+                RetainPtr annotationUnderMouse = annotationForRootViewPoint(event.position());
+                updateTrackedAnnotation(annotationUnderMouse.get());
+
```
Also this one

aprotyas reviewed (commented):
LGTM

sammygill committed:
[UnifiedPDF] Annotation tracking stops working when the mouse leaves the annotation.
https://bugs.webkit.org/show_bug.cgi?id=270117
rdar://123123950

Reviewed by Simon Fraser.

Currently annotation tracking begins in two scenarios:
1.) Mouse over of a texxt annotation
2.) Left click on other types of annotations

For the latter we finish tracking the annotation when the left mouse
button is released or if it is dragged off the tracked annotation. The
second case is incorrect, however, and we should contiue to keep track
of the annotation.

If the mouse is dragged off of the tracked annotation it should not be
highlighted anymore unless the mouse is dragged back onto it. If the
left mouse button is relased on the annotation then we want to perform
the associated action (e.g. activating a checkbox), but if it is
released off of the annotation then we should do nothing and just stop
tracking the annotation.

To accomplish the first part all we need to do is stop calling
finishTrackingAnnotation and instead call a new method:
updateTrackedAnnotation. All this new method does is update the
highlight of the tracked annotation but it does not modify the
annotation tracking state. For the second change, we need to pass in
the annotation under the mouse to finishTrackingAnnotation so the
annotation tracking state will only perform the associated action if
the annotation under the mouse is the annotation it is tracking.

One other small change that I made is to not change the cursor type when
the mouse is dragged off the annotation to make it consistent with
shipping.

* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::handleMouseEvent):
(WebKit::UnifiedPDFPlugin::updateTrackedAnnotation):
(WebKit::UnifiedPDFPlugin::finishTrackingAnnotation):
(WebKit::AnnotationTrackingState::finishAnnotationTracking):

Canonical link: https://commits.webkit.org/275362@main
